### PR TITLE
provider/aws: Support S3 bucket notification

### PR DESCRIPTION
**Pull request by kjmkznr**
Add [S3 bucket notification](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTnotification.html) resource.

The reason for creating a new resource is to avoid the circular reference.

```
resource "aws_sns_topic" "topic" {
    name = "terraform-test-topic"
        policy = <<POLICY
{
        "Version":"2012-10-17",
        "Statement":[{
                "Effect": "Allow",
                "Principal": {"AWS":"*"},
                "Action": "SNS:Publish",
                "Resource": "arn:aws:sns:*:*:terraform-test-topic",
                "Condition":{
                        "ArnLike":{"aws:SourceArn":"${aws_s3_bucket.bucket.arn}"}
                }
        }]
}
POLICY
}

resource "aws_s3_bucket" "bucket" {
        bucket = "tf-test-bucket-1928392189481395839258"
}

resource "aws_s3_bucket_notification" "bucket_notification" {
	bucket = "${aws_s3_bucket.bucket.id}"
	topic {
		topic_arn = "${aws_sns_topic.topic.arn}"
		events = ["s3:ObjectCreated:*"]
		filter_suffix = ".log"
	}
	queue {
		...
	}
	lambda_function {
		...
	}
}
```

### Acceptance test

```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSS3Bucket_Notification'                                                         
==> Checking that code complies with gofmt requirements...
/home/kjmkznr/go/bin/stringer
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSS3Bucket_Notification -timeout 120m
=== RUN   TestAccAWSS3Bucket_Notification
--- PASS: TestAccAWSS3Bucket_Notification (66.94s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/aws    66.945s
```

### CODE REVIEW

**Comment by stack72** [comment]
This may #4120

File: builtin/providers/aws/resource_aws_s3_bucket_notification.go [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+							Set:      schema.HashString,
+						},
+					},
+				},
+				Set: func(v interface{}) int {
+					var buf bytes.Buffer
+					m := v.(map[string]interface{})
+					buf.WriteString(fmt.Sprintf("%t-", m["filter_prefix"].(string)))
+					buf.WriteString(fmt.Sprintf("%t-", m["filter_suffix"].(string)))
+					return hashcode.String(buf.String())
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_s3_bucket_notification.go]: These `Set` functions do not account for all the elements in the set, so changing anything outside of `filter_prefix` or `filter_suffix` will not be recognized, e.g. changing to another SNS Topic. 

I believe it's safe to simply omit the `Set` func entirely here.

File: website/source/docs/providers/aws/r/s3_bucket_notification.html.markdown [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+resource "aws_sns_topic" "topic" {
+    name = "s3-event-notification-topic"
+    policy = <<POLICY
+{
+    "Version":"2012-10-17",
+    "Statement":[{
+        "Effect": "Allow",
+        "Principal": {"AWS":"*"},
+        "Action": "SNS:Publish",
+        "Resource": "arn:aws:sns:*:*:terraform-test-topic",
```
Comment (catsby) [inline website/source/docs/providers/aws/r/s3_bucket_notification.html.markdown]: This `terraform-test-topic` needs to match up to a valid SNS Topic name, correct? The example config does not complete successfully: 

```
* aws_s3_bucket_notification.bucket_notification: Error putting S3 notification configuration: InvalidArgument: Unable to validate the following destination configurations
```

Changing it to `s3-event-notification-topic` should work

File: builtin/providers/aws/resource_aws_sqs_queue.go [hunk 3]
```diff
@@ -177,6 +178,9 @@ func resourceAwsSqsQueueRead(d *schema.ResourceData, meta interface{}) error {
 					}
 					d.Set(iKey, value)
 				} else {
+					if iKey == "policy" {
+						*attrmap[oKey] = normalizeJson(*attrmap[oKey])
+					}
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_sqs_queue.go]: I'm curious why the changes here, do they directly contribute to the S3 notification resource, or is this something you encountered in development that needed patching? I'm curious if it needs testing

File: builtin/providers/aws/resource_aws_s3_bucket_notification.go (hunk 1 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_s3_bucket_notification.go]: Poking at this further, it seems that if omitted, the `id` will get populated by the API. This is troublesome because we'll then see a diff in future plans if we don't have that populated.

**Comment by catsby** [comment]
This looks really promising, I'm excited to have it in! We need to touch up some things first though as I noted above, please let me know if you have any questions about them

File: builtin/providers/aws/resource_aws_sqs_queue.go (hunk 3 above)
Comment (kjmkznr) [inline builtin/providers/aws/resource_aws_sqs_queue.go]: SQS Policy is always change every plan in current version.
If need split to other pull-request, I will do it.

```
resource "aws_s3_bucket" "bucket" {
   bucket = "tf-test-bucket-123456789" 
}

resource "aws_sqs_queue" "queue" {
    name = "terraform-test-queue"
        policy = <<POLICY
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": "*",
      "Action": "sqs:SendMessage",
          "Resource": "arn:aws:sqs:*:*:terraform-test-queue",
      "Condition": {
          "ArnEquals": { "aws:SourceArn": "${aws_s3_bucket.bucket.arn}" }
      }
    }
  ]
}
POLICY
}
```

```
$ terraform plan
...
~ aws_sqs_queue.queue
    policy: "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Principal\":\"*\",\"Action\":\"sqs:SendMessage\",\"Resource\":\"arn:aws:sqs:*:*:terraform-test-queue\",\"Condition\":{\"ArnEquals\":{\"aws:SourceArn\":\"arn:aws:s3:::tf-test-bucket-123456789\"}}}]}" => "{\n  \"Version\": \"2012-10-17\",\n  \"Statement\": [\n    {\n      \"Effect\": \"Allow\",\n      \"Principal\": \"*\",\n      \"Action\": \"sqs:SendMessage\",\n\t  \"Resource\": \"arn:aws:sqs:*:*:terraform-test-queue\",\n      \"Condition\": {\n          \"ArnEquals\": { \"aws:SourceArn\": \"arn:aws:s3:::tf-test-bucket-123456789\" }\n      }\n    }\n  ]\n}\n"
...
```

File: website/source/docs/providers/aws/r/s3_bucket_notification.html.markdown (hunk 2 above)
Comment (kjmkznr) [inline website/source/docs/providers/aws/r/s3_bucket_notification.html.markdown]: Fixed

File: builtin/providers/aws/resource_aws_sqs_queue.go (hunk 3 above)
Comment (kjmkznr) [inline builtin/providers/aws/resource_aws_sqs_queue.go]: I found opened pull-request, #4780.

And, I stopped use here document SQS policy.

**Comment by akerl** [comment]
+1 to this, just found it while trying to set up S3 -> lambda notifications

**Comment by akerl** [comment]
I'm digging into it a bit now, but when I use this with:

```
resource "aws_s3_bucket_notification" "cloudtrail_bucket_notification" {
    bucket = "${aws_s3_bucket.cloudtrail_bucket.id}"
    lambda_function {
        id = "notifylambda"
        lambda_function = "${aws_lambda_function.cloudtrail_alert_lambda.arn}"
        events = ["s3:ObjectCreated:*"]
    }
}
```

I get this:

```
* aws_s3_bucket_notification.cloudtrail_bucket_notification: Error putting S3 notification configuration: MalformedXML: The XML you provided was not well-formed or did not validate against our published schema
	status code: 400, request id: BLAH
```

**Comment by akerl** [comment]
Found the issue: if I add a filter_prefix of "*", it works:

```
resource "aws_s3_bucket_notification" "cloudtrail_bucket_notification" {
    bucket = "${aws_s3_bucket.cloudtrail_bucket.id}"
    lambda_function {
        id = "notifylambda"
        lambda_function = "${aws_lambda_function.cloudtrail_alert_lambda.arn}"
        events = ["s3:ObjectCreated:*"]
        filter_prefix = "*"
    }
}
```

I double checked what CloudTrail was catching for the error before:

```
    "requestParameters": {
        "bucketName": "BUCKETl",
        "NotificationConfiguration": {
            "CloudFunctionConfiguration": {
                "Filter": {
                    "S3Key": ""
                },
                "CloudFunction": "arn:aws:lambda:us-east-1:REDACTED:function:cloudtrail_alert_lambda",
                "Event": "s3:ObjectCreated:*",
                "Id": "notifylambda"
            }
        }
    },
```

So it looks like it should be tweaked to not provide a filter key to the API if no filter is provided to the resource definition

**Comment by akerl** [comment]
It turns out that "*" is taken as the literal *, not a wildcard. In my specific case, I just updated it to be AWSLogs, but that's not quite the global workaround I'd thought it would be.

File: website/source/docs/providers/aws/r/s3_bucket_notification.html.markdown [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+		events = ["s3:ObjectCreated:*"]
+		filter_suffix = ".log"
+	}
+}
+```
+
+### Add notification configuration to Lambda Function
+
+```
+resource "aws_iam_role" "iam_for_lambda" {
```
Comment (kjmkznr) [inline website/source/docs/providers/aws/r/s3_bucket_notification.html.markdown]: I can't reproduce fail put notification configuration.
Please tell me what kind of AWS returned error code.
Following error?

* Error Code: InvalidArgument
* Error Message: Unable to validate the following destination configurations

File: builtin/providers/aws/resource_aws_s3_bucket_notification.go (hunk 1 above)
Comment (kjmkznr) [inline builtin/providers/aws/resource_aws_s3_bucket_notification.go]: @catsby 
I think `id` parameter is good to change to auto generate by terraform when empty like AutoScaling Group name.
What do you think?

**Comment by kjmkznr** [comment]
@akerl Thanks find bug.
Fixed failed when not specific both `filter_prefix` and `filter_suffix` parameters.
Please try again.

**Comment by curena** [comment]
This is looking really nice. I hope it makes it into 0.6.14.

File: builtin/providers/aws/resource_aws_s3_bucket_notification.go (hunk 1 above)
Comment (kjmkznr) [inline builtin/providers/aws/resource_aws_s3_bucket_notification.go]: I changed to auto generate `id` when not specific `id` parameter.

**Comment by curena** [comment]
What's the status on this PR? Will this make it to the next release?

**Comment by stack72** [comment]
This looks now good that the initial fixes were made. I am going to merge and squash the commits

test results are as follows:

```
TF_LOG=1 make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSS3Bucket_Notification' 2>~/tf.log
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSS3Bucket_Notification -timeout 120m
=== RUN   TestAccAWSS3Bucket_Notification
--- PASS: TestAccAWSS3Bucket_Notification (67.51s)
=== RUN   TestAccAWSS3Bucket_NotificationWithoutFilter
--- PASS: TestAccAWSS3Bucket_NotificationWithoutFilter (28.43s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	95.958s
```
